src/components/legend.ts

```diff
@@ -1,4 +1,233 @@
+      const indexOfColumn = this._columns.indexOf(column);
+      const columnXOffset = this._columns.slice(0, indexOfColumn)
+        .reduce((sum, column) => sum + column.getWidth(), 0);
+
+      return {
+        topLeft: { x: columnXOffset, y: 0 },
+        bottomRight: {
+          x: columnXOffset + column.getWidth(),
+          y: column.getHeight(),
+        }
+      };
+    }
+
+    public getColumns() {
+      return this._columns;
+    }
+
+    /**
+     * Returns the height of the row, including the bottomPadding.
+     * @return {number} height
+     */
+    public getHeight() {
+      return Utils.Math.max(this._columns.map((column) => column.getHeight()), 0)
+        + this._bottomPadding;
+    }
+
+    /**
+     * Padding applied below the row. Affects the spacing between rows. Defaults to 0.
+     * @param {bottomPadding} number
+     */
+    public setBottomPadding(bottomPadding: number) {
+      this._bottomPadding = bottomPadding;
+    }
+
+    /**
+     * Sets the maximum allowable width of this column.
+     * @param {number} maxWidth
+     */
+    public setMaxWidth(maxWidth: number) {
+      this._maxWidth = maxWidth;
+    }
+
+    /**
+     * Returns the current width of the row constrained by maxWidth, if set.
+     * @returns {number} width
+     */
+    public getWidth() {
+      return Math.min(
+        this._columns.reduce((sum, column) => sum + column.getWidth(), 0),
+        this._maxWidth
+      );
+    }
+
+    /**
+     * Returns the remaining width available in the row based on the maximum
+     * width of this row.
+     * @returns {number} widthRemaining
+     */
+    public getWidthAvailable() {
+      const widthConsumed = this.getWidth();
+      return Math.max(this._maxWidth - widthConsumed, 0);
+    }
+  }
+
+  /**
+   * Stores LegendRows. Useful for calculating and maintaining
+   * positioning information about the Legend.
+   */
+  class LegendTable {
+    constructor(
+        private _maxWidth = Infinity,
+        private _maxHeight = Infinity,
+        private _padding = 0,
+        private _rows: LegendRow[] = []) {}
+
+    public addRow(row: LegendRow) {
+      row.setMaxWidth(this._maxWidth - this._padding * 2);
+      this._rows.push(row);
+    }
+
+    public forEach(callback: (row: LegendRow, index?: number) => void) {
+      this._rows.forEach(callback);
+    }
+
+    /**
+     * Returns the bounds relative to the parent and siblings of the row or
+     * column. If column is not specified, bounds will be calculate for the row only.
+     * If the column is specified, the bounds will be calculate for the column only.
+     *
+     * @param {LegendRow} row The row to calculate bounds
+     * @param {LegendColumn} column The column to calculate bounds.
+     * @returns {Bounds}
+     */
+    public getBounds(row: LegendRow, column?: LegendColumn<any>) {
+      const indexOfRow = this._rows.indexOf(row);
+      const rowYOffset = this._rows.slice(0, indexOfRow)
```

Comment:
we're trading a memory copy for iteration, which is going to be slower in javascript than just scanning the whole array. Still, it seems like the faster thing would just be to do normal for-loop iteration with an early `break`.